src/cljx/schema/core.cljx

```diff
@@ -473,47 +422,71 @@
 
 (clojure.core/defrecord Either [schemas]
   Schema
-  (walker [this]
-          (let [sub-walkers (mapv subschema-walker schemas)]
-            (clojure.core/fn [x]
-              (loop [sub-walkers (seq sub-walkers)]
-                (if-not sub-walkers
-                  (macros/validation-error
-                   this x
-                   (list 'some (list 'check '% (utils/value-name x)) 'schemas))
-                  (let [res ((first sub-walkers) x)]
-                    (if-not (utils/error? res)
-                      res
-                      (recur (next sub-walkers)))))))))
+  (spec [this]
+    (variant/variant-spec
+     spec/+no-precondition+
+     (for [s schemas]
+       {:guard (complement (checker s))
+        :schema s})
+     #(list 'matches-some-either-clause? %)))
   (explain [this] (cons 'either (map explain schemas))))
 
 (clojure.core/defn either
-  "A value that must satisfy at least one schema in schemas."
+  "A value that must satisfy at least one schema in schemas.
+   Note that `either` does not work properly with coercion
+
+   WARNING: either does not work with coercion.  It is also slow and gives
+   bad error messages.  Please consider using `conditional` and friends
+   instead; they are more efficient, provide better error messages,
+   and work with coercion."
   [& schemas]
   (Either. schemas))
 
 
+;;; constrained (matches precondition and schema)
+
+(clojure.core/defrecord Constrained [schema pred fn-name]
```

Comment:
:sheep: is this the same as `conditional` with only 1 pred-schema pair and different error messages? not sure why it would be preferred to conditional?